### PR TITLE
feat(Tag): Implement manage tags dialog

### PR DESCRIPTION
**Pull request by geoffreykwan**
## Changes
- Implemented manage tags dialog
- Tag text can now be edited

## Test
- Test with https://github.com/WISE-Community/WISE-API/pull/261
- In the Class Schedule view, select some units by clicking on their checkbox
- Click the "Apply Tags" button
- Click the "Manage Tags" menu item
- Change the text for a tag. The tag text should immediately update in the "Apply tags" menu but not in the unit cards. I'll fix immediately updating in the unit cards in another PR.

### CODE REVIEW

File: src/app/teacher/apply-tags-button/apply-tags-button.component.ts [hunk 1]
```diff
@@ -21,6 +23,10 @@ export class ApplyTagsButtonComponent implements OnInit {
       }
       this.tags = tags;
     });
+    this.tagService.tagChanged$.subscribe((tagThatChanged: Tag) => {
+      const tag = this.tags.find((t: Tag) => t.id === tagThatChanged.id);
+      tag.text = tagThatChanged.text;
+    });
```
Comment (hirokiterashima) [inline src/app/teacher/apply-tags-button/apply-tags-button.component.ts]: Does this need to be unsubscribed in ```ngOnDestroy()```? Also, would ```updatedTag``` be a better name here?

File: src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+})
+export class ManageTagsDialogComponent implements OnInit {
+  inputChanged: Subject<string> = new Subject<string>();
+  inputChangedSubscription: Subscription;
+  protected tags: Tag[] = [];
+
+  constructor(private tagService: TagService) {}
+
+  ngOnInit(): void {
+    this.tagService.retrieveUserTags().subscribe((tags: Tag[]) => {
```
Comment (hirokiterashima) [inline src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.ts]: unsubscribe in onDestroy()?

File: src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    FlexLayoutModule,
+    FormsModule,
+    MatButtonModule,
+    MatDialogModule,
+    MatInputModule,
+    MatFormFieldModule
+  ]
+})
+export class ManageTagsDialogComponent implements OnInit {
+  inputChanged: Subject<string> = new Subject<string>();
```
Comment (hirokiterashima) [inline src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.ts]: protected?

File: src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  inputChangedSubscription: Subscription;
```
Comment (hirokiterashima) [inline src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.ts]: private?

File: src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  constructor(private tagService: TagService) {}
+
+  ngOnInit(): void {
+    this.tagService.retrieveUserTags().subscribe((tags: Tag[]) => {
+      this.tags = tags;
+    });
+    this.inputChangedSubscription = this.inputChanged
+      .pipe(debounceTime(1000), distinctUntilChanged())
+      .subscribe(({ tag }: any) => {
+        this.updateTag(tag);
```
Comment (hirokiterashima) [inline src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.ts]: Call service function directly and remove need to define a new function?
```suggestion
        this.tagService.updateTag(tag);
```

File: src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.html [hunk 6]
```diff
@@ -0,0 +1,18 @@
+<h2 mat-dialog-title i18n>Manage Tags</h2>
+<mat-dialog-content>
+  <div fxLayout="column" fxLayoutGap="12px">
+    <div *ngFor="let tag of tags" fxLayout="row" fxLayoutAlign="start center">
+      <mat-form-field subscriptSizing="dynamic" fxFlex>
+        <input
+          matInput
+          type="text"
+          [(ngModel)]="tag.text"
+          (ngModelChange)="inputChanged.next({ tag: tag, text: $event })"
```
Comment (hirokiterashima) [inline src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.html:10]: ```ManageTagsDialogComponent.inputChanged``` has type ```Subject<string>```, but here we're passing in an object. What should it be?

File: src/assets/wise5/services/tagService.ts [hunk 7]
```diff
@@ -100,4 +102,10 @@ export class TagService {
     }
     return this.http.delete(`/api/projects/tag/${tag.text}`, { params: params }).subscribe();
   }
+
+  updateTag(tag: Tag): Subscription {
+    return this.http.put<Tag>(`/api/user/tag/${tag.id}`, tag).subscribe((tag) => {
+      this.tagChangedSource.next(tag);
+    });
+  }
```
Comment (hirokiterashima) [inline src/assets/wise5/services/tagService.ts]: Do we need to return a Subscription? It looks like we don't do anything with it?
```suggestion
  updateTag(tag: Tag): void {
    this.http.put<Tag>(`/api/user/tag/${tag.id}`, tag).subscribe((tag) => {
      this.tagChangedSource.next(tag);
    });
  }
```

The other option is to return an Observable<Tag> and subscribe in the component.
```
  updateTag(tag: Tag): Observable<Tag> {
    return this.http.put<Tag>(`/api/user/tag/${tag.id}`, tag).pipe(
      tap((tag) => {
        this.tagChangedSource.next(tag);
      })
    );
  }
```

**Review by hirokiterashima — CHANGES_REQUESTED** [review]
Functionality works as described. I have a few comments:

1. We discussed using a snack bar to indicate that the change has been saved.
2. I added some code change suggestions inline.

I'm also wondering if we should split up the TagService to StudentTagService and UserTagService. It looks like there are two different kinds of tags right now, and handling them both in TagService is confusing. This refactoring can be done in a separate PR. Let's discuss how best to proceed.

**Comment by geoffreykwan** [comment]
Ok, I've made the changes.

**Review by hirokiterashima — APPROVED** [review]
LGTM

In ManageTagsDialogComponent, 
```protected inputChanged: Subject<string> = new Subject<any>();```
should be 
```protected inputChanged: Subject<any> = new Subject<any>();```
?

**Pull request merged**